proposed/0000-Unified-structured-data.md

```diff
@@ -0,0 +1,128 @@
+on the network and how it is handled.
+
+###Data storage and retrieval
+
+ImmutableData is fixed self validating non mutable chunks. These require StructuredData types to manipulate information. These structured Data types may then create a global application acting on a key value store with very high degrees of availability and security (i.e. create network scale apps). Such apps could easily include medical condition analysis linked with genomic and proteomic sequencing to advance health based knowledge on a global scale. This proposal allows such systems to certainly be prototyped and tested with a high degree of flexibility.
+
+###New protocols
+
+As these data types are now self validating and may contain different information, such as new protocols, `rdf`/`owl` data types, the limit of new data types and ability to link such data is extremely scalable. Such protocols could indeed easily encompass token based systems (a form of 'crypto-currency'), linked data, natural language learning databases, pre-compilation units, distributed version control systems (git like) etc.
+
+###Compute
+
+Such a scheme would allow global computation types, possibly a Domain Specific Language (DSL) would define operator types to allow combination of functions. These could be made monotonic and allow out of order processing of programs (disorderly programming) which in itself presents an area that may prove to be well aligned with decentralised 'intelligence' efforts. Linked with 'zk-snarks' to alleviate any 'halting problem' type issues then a global Turing complete programming environment that optionally acts on semantic ('owl' / 'json-ld' etc.) data is a possible.
+
+##Expected outcome
+
+
+It is expected removing Transaction Managers from network will reduce complexity, code and increase security on the network, whilst allowing a greater degree of flexibility. These types now allow the users of this network to create their own data types and structures to be securely managed. This is hoped to allow many new types of application to exist.
+
+# Detailed design
+
+The design entails reducing all StructuredData types to a single type, therefore it should be able to be recognised by the network as StructuredData and all such sub-types handled exactly in the same manner. 
+
+##StructuredData
+
+```
+struct StructuredData {
+tag_type : TagType, // 8 Bytes ?
+identifier : NameType // 64Bytes (i.e. SHA512 Hash)
+data : mut Vec<u8>, // in many cases this is encrypted
+owner_keys : mut vec<crypto::sign::PublicKey> // n * 32 Bytes (where n is number of owners)
+version : mut u64, // incrementing (deterministic) version number
+previous_owner_keys : mut vec<crypto::sign::PublicKey> // n * 32 Bytes (where n is number of
+owners) only required when owners change 
+signature : mut Vec<Signature> // signs the `mut` fields above // 32 bytes (using e25519 sig)
+}
+```
+
```

Comment:
I'd propose to change this to:

```rust
/// Top-level type: a representation of "Structured Data".
pub struct Data {
    /// Immutable attributes which apply to the entire `Data` instance.
    fixed_attributes: FixedAttributes,
    /// Attributes which apply to the entire `Data` instance, but which can be changed with proper
    /// authorisation.
    mutable_attributes: MutableAttributes,
    /// The most recent (which could encompass all) versions of the `Data` instance.  Cannot be
    /// empty.
    versions: Vec<Version>,
}

/// Attributes of the `Data` which can never change once initially set.  These define the identity,
/// type and some of the rules the network will employ for handling the `Data`.  It can also hold
/// arbitrary data which will likely be meaningless to the network.
pub struct FixedAttributes {
    /// Identifier of the `Data` type.
    type_tag: u64,
    /// Identity of the piece of `Data`.
    id: NameType,
    /// Maximum number of versions allowed.
    max_versions: u64,
    /// Number of versions to retain when archiving a "full" piece of `Data` (minimum value of 1).
    min_retained_count: u8,
    /// Arbitrary, immutable, `Data`-wide information.  May be empty.
    data: Vec<u8>,
}

/// A representation of an owner's public key and the bias which should be given to that key when
/// a mutating request is received by the network.
pub struct KeyAndWeight {
    /// Owner's public key.
    key: sign::PublicKey,
    /// Bias given to this public key (minimum value of 1).
    weight: u64
}

/// Attributes of the `Data` which can be changed via a properly-authorised request to the network.
/// These define the current owner's public keys, further rules the network will employ for handling
/// the `Data` and also arbitrary data which will likely be meaningless to the network.
pub struct MutableAttributes {
    /// Current owner or owners' public keys.  Cannot be empty.
    owner_keys: Vec<KeyAndWeight>,
    /// Minimum total weight of signatories' keys to allow a mutation of the piece of `Data` (at
    /// least one signature will be required regardless of this minimum).
    min_weight_for_consensus: u64,
    /// Coarse-grained expiry date around which time the piece of `Data` will be removed from the
    /// network.
    expiry_date: time::Tm,
    /// Arbitrary, mutable, `Data`-wide information.  May be empty.
    data: Vec<u8>,
}

/// A representation of a single version.  The `index` allows provision of strict total ordering of
/// the `Version`s.  It can also hold arbitrary data specific to that particular `Version`, e.g.
/// encrypted content or the name of a piece of "Immutable Data".
pub struct Version {
    /// Sequential number to provide strict total order of versions.
    index: u64,
    /// Arbitrary, version-specific information.  May be empty.
    data: Vec<u8>,
}
```